Add new regex config parameter version to blacklisted files and excluded directories

mmattel opened the pull request:
This PR is based on the work of @tdaget at PR https://github.com/owncloud/core/pull/36331 (closed)
The referenced PR is now closed because of rebasing issues and CI will never get green there.
I improved the original work a bit.

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
This PR enhances the existing config parameters `blacklisted_files` and `excluded_directories` by a new regex parameter version `blacklisted_files_regex` and `excluded_directories_regex`. 

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- No issue related

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Sometimes it is necessary to exclude/blacklist files or directories to be written/renamed/scanned on a more complex parameter than a static version as we have right now. This PR allows eg. to prevent uploading Microsoft Outlook .pst files or creating / renaming / scanning directories based on a pattern defined by a regex. The existing parameters are not touched.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Running local with various scenarios successfully.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: Not necessary as doc is part of config.sample.php 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

tdaget reviewed (changes requested):
An error is located in each preg_match call, regex must be surrounded by `\`. Plus, as regex match is case insensitive an `i` must be added to the end of the string.

In lib/private/Files/Filesystem.php:
It might be: `\preg_match('/'.$item.'/i', $path_part)`

tdaget reviewed (changes requested):
You add a comment for regex variables :
 * INFO: Compare pairs are converted to lowercase by default.

I'm ok for regular variables, for regex the comment might be adapted as things are not simple as a lowercase string.

Maybe :
 * INFO: regex matching is forced case insensitive.

mmattel commented:
@tdaget thanks for the comments, both fixed

mmattel commented:
@phil-davis can you do me a favour and take a look why CI fails.
I have rebased on latest master (on and on...) and it never gets green.

phil-davis commented:
The unit tests fail:
```
There were 3 failures:

1) Test\Files\FilesystemTest::testIsFileBlacklistedRegex with data set #6 ('/etc/dummy/machin.chose', array('.*\.pst$', '.*dummy.*', '^sample.*'), false)
Failed asserting that true matches expected false.

/drone/src/tests/lib/Files/FilesystemTest.php:299

2) Test\Files\FilesystemTest::testIsExcludedRegex with data set #8 ('/foo/_Thomas/Files', array('.*backup.*', 'Thomas.*'), false)
Failed asserting that true matches expected false.

/drone/src/tests/lib/Files/FilesystemTest.php:347

3) Test\Files\FilesystemTest::testIsExcludedRegex with data set #10 ('/foo/_ThomasFoo/Files', array('.*backup.*', 'Thomas.*'), false)
Failed asserting that true matches expected false.

/drone/src/tests/lib/Files/FilesystemTest.php:347
```

mmattel commented:
Tests fixed (locally tested), rebased and pushed, CI is green :smile:
@phil-davis can you do a review pls

phil-davis commented on config/config.sample.php:
```diff
@@ -1204,24 +1204,57 @@
+ * INFO: regex matching is forced case insensitive.
+ * INFO: Leading and trailing slashes `/` must not be manually added.
+ * WARNING: USE THIS ONLY IF YOU KNOW WHAT YOU ARE DOING.
+ * WARNING: COMPLEX REGEX MAY SIGNIFICANT IMPACT YOUR SCANNING PERFORMANCE.
```
```suggestion
 * WARNING: COMPLEX REGEX MAY SIGNIFICANTLY IMPACT YOUR SCANNING PERFORMANCE.
```

phil-davis commented on config/config.sample.php:
```diff
@@ -1204,24 +1204,57 @@
+ * INFO: regex matching is forced case insensitive.
+ * INFO: Leading and trailing slashes `/` must not be manually added.
+ * WARNING: USE THIS ONLY IF YOU KNOW WHAT YOU ARE DOING.
+ * WARNING: COMPLEX REGEX MAY SIGNIFICANT IMPACT YOUR SCANNING PERFORMANCE
```
```suggestion
 * WARNING: COMPLEX REGEX MAY SIGNIFICANTLY IMPACT YOUR SCANNING PERFORMANCE
```

phil-davis commented on config/config.sample.php:
```diff
@@ -1204,24 +1204,57 @@
+ */
+'excluded_directories_regex' => [
+	'.*backup.*',	// exclude folders with name containing the string 'backup'
+	'thomas.*',		// exclude folders with name beginning by 'thomas'
```
```suggestion
	'thomas.*',		// exclude folders with name beginning with 'thomas'
```

phil-davis commented on config/config.sample.php:
```diff
@@ -1204,24 +1204,57 @@
+'blacklisted_files_regex' => [
+	'.*\.pst$',		// disallow *.pst (MS Outlook) file
+	'.*dummy.*',	// disallow file with name contaning the string 'dummy'
+	'^sample.*',	// disallow file with name beginning by 'sample'
```
```suggestion
	'^sample.*',	// disallow file with name beginning with 'sample'
```

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,165 @@ public static function isIgnoredDir($dir) {
+	}
+
+	/**
+	 * Regex validy check.
```
```suggestion
	 * Regex validity check.
```

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,165 @@ public static function isIgnoredDir($dir) {
+	 * @param string $callerMessage message from the calling function to identify where it is coming from
+	 * @return boolean
+	 */
+	private static function regexValidyCheck($regexToBeChecked, $callerMessage='') {
```
```suggestion
	private static function regexValidityCheck($regexToBeChecked, $callerMessage='') {
```

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,165 @@ public static function isIgnoredDir($dir) {
+	 * Manage regex against Folder or File.
+	 * @param array $regexList Regex array as defined in Config file
+	 * @param array $path Folder path or File contained in an array
+	 * @param string $callerMessage Message from the calling function to identify where a possible error it is coming from
```
```suggestion
	 * @param string $callerMessage Message from the calling function to identify where a possible error is coming from
```

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,165 @@ public static function isIgnoredDir($dir) {
+			} else {
+				$newItem = $newItem . 'i';
+			}
+			@\preg_match($newItem, null);                         // regex validy check
```
```suggestion
			@\preg_match($newItem, null);                         // regex validity check
```

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,165 @@ public static function isIgnoredDir($dir) {
+				$newItem = $newItem . 'i';
+			}
+			@\preg_match($newItem, null);                         // regex validy check
+			if (self::regexValidyCheck($item, $callerMessage)) {  // check if regex error occur
```
```suggestion
			if (self::regexValidyCheck($item, $callerMessage)) {  // check if regex error occurred
```

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,165 @@ public static function isIgnoredDir($dir) {
+			if (self::regexValidyCheck($item, $callerMessage)) {  // check if regex error occur
+				foreach ($path as $path_part) {                   // foreach folder or file in given path
+					if (@\preg_match($newItem, $path_part)        // regex match item
+						&& self::regexValidyCheck($item,
```
```suggestion
						&& self::regexValidityCheck($item,
```

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,165 @@ public static function isIgnoredDir($dir) {
+				foreach ($path as $path_part) {                   // foreach folder or file in given path
+					if (@\preg_match($newItem, $path_part)        // regex match item
+						&& self::regexValidyCheck($item,
+								'Checked string: '.$path_part)) { // check if regex error occur
```
```suggestion
								'Checked string: '.$path_part)) { // check if regex error occurred
```

phil-davis reviewed (commented):
Just nits about spelling etc.
Somebody in product management needs to decide if this feature should be added.

mmattel commented:
@micbar @pmaier1 adding you according the suggestion of @phil-davis 
This PR is a great and useful feature, adding good functionality, not breaking an exiting config.
Your review / approval is welcomed.

mmattel commented:
@phil-davis CI did not started at all after pushing a rebase this morning? Can you help?

phil-davis commented:
> @phil-davis CI did not started at all after pushing a rebase this morning? Can you help?

drone has a huge backlog today. It seems to have missed some triggers from PRs. Rebase again.

mmattel commented:
@micbar mind reviewing, CI is green

mmattel commented:
squashed and rebased

micbar commented on changelog/unreleased/36325:
```diff
@@ -0,0 +1,7 @@
+Enhancement: Adding a regex version to config option of blacklisted_files and excluded_directories
+
```
The title is limited to 80 chars.

mmattel committed:
Regex version for blacklisted_files and excluded_directories

mmattel commented:
Ping?

phil-davis commented:
@micbar IMO this code is OK and the feature works. The next question is will it be accepted into core?

If so, then we can schedule adding a few acceptance test scenarios...

mmattel commented:
@micbar ping?

tdaget commented:
+1 ...

pmaier1 commented:
> @micbar @pmaier1 adding you according the suggestion of @phil-davis
This PR is a great and useful feature, adding good functionality, not breaking an exiting config.
Your review / approval is welcomed.

Great work! Thanks for the contribution. We'd accept it and take it over to the product. Still, we would highly appreciate if you could complete the checklist in the PR template and add the relevant acceptance tests. I think @phil-davis and @micbar can support you with this.

Does this work for you?

phil-davis commented:
This is an enhancement, so would need to go in a minor release (according to semver "rules)
So that means either add tests ASAP and get it merged for 10.4, or it will end up waiting "however long" until we do a 10.5 
@micbar comment?

micbar commented:
@phil-davis 10.4 is out of scope here. We need to have qa on this first.

@mmattel We have a good general explanation of acceptance tests in our DEV docs.
https://doc.owncloud.com/server/developer_manual/testing/acceptance-tests.html

We can use your contribution. We need to handle it like many other open source projects.
We can decide to merge it but we cannot maintain in on ourselves. So we need your commitment in maintaining this and help testing it.

@phil-davis can assist.

mmattel commented:
@micbar @pmaier1 as you know me, I stand for the code I am contributing - since day 1 of my participation. The code always works as expected and wanted and if there is any upcoming issue, I commit maintaining it (btw, this was never necessary in all the years :smile:)

@phil-davis I would appreciate your assistance for the tests

phil-davis commented:
@mmattel I am offline all tomorrow, so I can look at the weekend or Monday.
Without having looked closely, there should be acceptance tests for the existing blacklist options and hopefully it is "just" a matter of copy-paste and edit some new scenarios that test the new options using interesting regex.

cdamken commented:
Is this already tested with sync-clients?

mmattel commented:
@cdamken this is a (regex) extension of the existing blacklist/excluded functionality which is present since ages

phil-davis committed:
Code review of blacklist and excluded files regex

phil-davis commented:
I rebased and added a commit with various code-review adjustments.
Now I will add some acceptance tests.

phil-davis committed:
Add examples of excluded/blacklisted regex in config.sample.php

phil-davis commented on config/config.sample.php:
```diff
@@ -1291,6 +1303,23 @@
+ *
+ * WARNING: USE THIS ONLY IF YOU KNOW WHAT YOU ARE DOING.
+ */
+'excluded_directories' => [
```
```suggestion
'excluded_directories_regex' => [
```

phil-davis committed:
tests for blacklisted_files_regex and excluded_directories_regex

upload tests for blacklisted_files_regex and excluded_directories_regex

rename tests for blacklisted_files_regex and excluded_directories_regex

move async tests for blacklisted_files_regex and excluded_directories_regex

move folder tests for blacklisted_files_regex and excluded_directories_regex

webUI tests for blacklisted_files_regex and excluded_directories_regex

uploadFileAsyncUsingNewChunking tests for blacklisted_files_regex and excluded_directories_regex

uploadFileUsingNew-and-OldChunking tests for blacklisted_files_regex and excluded_directories_regex

micbar commented:
@phil-davis Wow! Impressive amount of Tests!

phil-davis commented:
I cherry-picked from https://github.com/owncloud/core/pull/36643 where I added some upload tests for the existing options, so it looks more than it is.

I still need to add tests for webDAV move/delete/copy etc. I have those already in core master now for the existing features, so in theory I just copy-paste the scnearios and adjust them to have some example regex. Hope to do that in the next hours...

phil-davis committed:
Adjust tests to demonstrate that the regex match is case-insensitive

phil-davis commented:
@micbar I have completed acceptance tests for this. The new regex features work.

Somebody could give a technical review of the source code changes.

@individual-it can do a review of the acceptance tests.

If this is approved, then we could also cherry-pick this back into release-10.4 branch so it can get released.

In making tests for the existing blacklist and excluded directories features, I discovered https://github.com/owncloud/core/issues/36645 - in "old" "v1" chunking uploads the blocked uploads return a 507 (rather than 403). The uploads are blocked - so that existing issue is not so important. Similar behaviour happens in the new regex features.

individual-it reviewed (commented):
tests look good
we could move them all over to a separate feature file though

phil-davis commented:
I created issue #36654 for the reorganisation of feature files, because that impacts more than just the scenarios in this PR

phil-davis commented:
Rebased and resolved conflict

phil-davis committed:
skip new regex blacklist tests on 10.3

karakayasemi commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,163 @@ public static function isIgnoredDir($dir) {
+			}
+			@\preg_match($newItem, null);
+			if (self::regexValidityCheck($item, $callerMessage)) {
+				foreach ($path as $path_part) {
```
please use camel case for variable names.

karakayasemi commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,163 @@ public static function isIgnoredDir($dir) {
+		// force blacklist/exclude arraylist/arrayRegex for unit tests
 		if ($ed) {
-			$excluded = $ed;
+			$exclude_folders = $ed;
```
please use camel case for variable names.

phil-davis replied:
PR incoming to fix the var naming.

karakayasemi commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,163 @@ public static function isIgnoredDir($dir) {
+		// important, because if you define an empty config value, '.htaccess' will not be added...
+		// prevents misuse with a fake config value
+		$blacklist_files[] = '.htaccess';
+		$blacklist_files = \array_unique($blacklist_files);
```
To prevent code duplication, we can move these assignments to a private function. But, I am not sure is it worth. It is up to writer.

karakayasemi commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,163 @@ public static function isIgnoredDir($dir) {
-			$match = \array_intersect($path_parts, $excluded);
-			if ($match) {
+
+		// force that the last element (possibly the filename) is an array
```
This line is confusing, the last element of `$path_parts` is string, not an array. As far as understand, what is meant here, the last element of `$path_parts` should be holded on an array to easy comparison or I am misunderstanding.

phil-davis replied:
I have rewritten the comment a bit. PR incoming.

karakayasemi commented on config/config.sample.php:
```diff
@@ -1291,6 +1303,23 @@
+ *
+ * WARNING: USE THIS ONLY IF YOU KNOW WHAT YOU ARE DOING.
+ */
+'excluded_directories_regex' => [
```
Should not this only apply to directories? As far as I see, we are also applying it to files.

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,163 @@ public static function isIgnoredDir($dir) {
-	* @param array $ed
+	 *
+	* @param string $FileOrDir is an array describing full folder path or full filename
+	* @param array $ed is only used in conjunction with unit tests as we handover here the blacklisted / excluded
```
I will call it `$excluded`
Note: it was already called `$ed` before this PR

mmattel replied:
It was named `$ed` because of the currently not available constructor in tests and not to get confused by all the other `$excluded` variables below. Maybe better to give it a more meaningful name like `$excludedTest` ... ?

phil-davis commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,49 +625,163 @@ public static function isIgnoredDir($dir) {
+		$blacklist_array[] = \end($path_parts);
+
+		// first, check if the folder is excluded
+		if (isset($exclude_folders)) {
```
Hmmm - I am not even sure why this test is here - there is already code above that always sets `$exclude_folders` - so the code inside this `if` will always be run.
I will remove the unneeded if in all 4 places.

phil-davis commented:
See #36694 and #36695 for the changes to address the review comments
